tx-compiler/src/compiler.rs

```diff
@@ -0,0 +1,299 @@
+            .with_library(&StdLibrary::default())
+            .expect("library is well formed")
+            .with_kernel_module(tx_module)
+            .expect("kernel is well formed");
+        TransactionComplier {
+            assembler,
+            account_procedures: HashMap::default(),
+            prologue: TransactionKernel::prologue(),
+            epilogue: TransactionKernel::epilogue(),
+            note_setup: TransactionKernel::note_setup(),
+            tx_kernel,
+        }
+    }
+
+    /// Compiles the provided module into [AccountCode] and associates the resulting procedures
+    /// with the specified account ID.
+    pub fn load_account(
+        &mut self,
+        account_id: AccountId,
+        account_code: ModuleAst,
+    ) -> Result<AccountCode, TransactionError> {
+        let account_code = AccountCode::new(account_code, account_id, &mut self.assembler)
+            .map_err(TransactionError::LoadAccountFailed)?;
+        self.account_procedures.insert(account_id, account_code.procedures().to_vec());
+        Ok(account_code)
+    }
+
+    /// Loads the provided account interface (vector of procedure digests) into the `account_procedures` map.
+    /// Returns the old account interface if it previously existed.
+    pub fn load_account_interface(
+        &mut self,
+        account_id: AccountId,
+        procedures: Vec<Digest>,
+    ) -> Option<Vec<Digest>> {
+        self.account_procedures.insert(account_id, procedures)
+    }
+
+    /// Compiles the provided program into the [NoteScript] and checks (to the extent possible)
+    /// if a note could be executed against an account with the specified interface.
+    pub fn compile_note_script(
+        &mut self,
+        note_script_ast: ProgramAst,
+        // TODO: Should this be optional - what if we don't know the target account interface.
+        target_account_proc: Vec<NoteTarget>,
+    ) -> Result<NoteScript, TransactionError> {
+        let (note_script, code_block) = NoteScript::new(
+            note_script_ast,
+            &mut self.assembler,
+            &mut AssemblyContext::new(AssemblyContextType::Program),
+        )
+        .map_err(|_| TransactionError::CompileNoteScriptFailed)?;
+
+        let branches = collect_call_branches(&code_block);
+        for note_target in target_account_proc.into_iter() {
+            verify_note_account_compatibility(
+                note_script.hash(),
+                &branches,
+                &self.get_target_interface(note_target)?,
+            )?;
+        }
+
+        Ok(note_script)
+    }
+
+    /// Returns the account interface asscoiated with the provided [NoteTarget].
+    ///
+    /// # Errors
+    /// - If the account interface associated with the [AccountId] provided as a target can not be
+    ///   found in the `account_procedures` map.
+    fn get_target_interface(&self, target: NoteTarget) -> Result<Vec<Digest>, TransactionError> {
+        match target {
```

Comment:
Would changing the signature to something like:
```Rust
fn get_target_interface<'a>(&self, target: &'a NoteTarget) -> Result<&'a [Digest], TransactionError>
```
not work here? (to try to avoid cloning).

Reply:
I wasn't able to get this working.  The problem is the lifetime of the return type is different depending on the `target` enum argument.  If `target == NoteTarget::Procedures` then the return type has a lifetime equal to target.  If `target == NoteTarget::AccountId` then the return type has a lifetime equal to self.